### ISSUE TITLE
Return  value of Push.create is undefined for the first invocation

### ISSUE THREAD

**Issue by venkatramachandran** [Nickersoft/push.js#13]
The first time a notification is created using `Push.create`, the return value is `undefined`. This is reproducible in the `example.html` provided. 
Because of this, calling `close()` on the notification fails.

**Comment by Nickersoft** [comment]
You say for the first invocation... upon calling it in future requests, is it not undefined?

**Comment by venkatramachandran** [comment]
That is correct.

**Comment by venkatramachandran** [comment]
```
/* Request permission if it isn't granted */
            if (!self.Permission.has()) {
                self.Permission.request(function () {
                    return create_callback(title, options);
                });
            } else {
                return create_callback(title, options);
            }
```
When the flow has to request the permission, the return value is `undefined` to the caller.

**Comment by sarbbottam** [comment]
I guess it would better to `promisify` and return a `promise` ...

**Comment by Nickersoft** [comment]
@venkatramachandran Oh I see what you mean... guess that was an oversight. I can just change the scope of that return and it should work just fine, unless @sarbbottam, is there any definitive advantage to using a promise? I'm not too familiar with them. Could you elaborate?

**Comment by sarbbottam** [comment]
```js
if (!self.Permission.has()) {
    // self.Permission.request() is a async operation
    // whenever self.Permission.request() is invoked
    // self.create will return undefined
    // for sure you don't want to make self.Permission.request() synchronous
    // so, if you like to return a handle to the user 
    // I guess you have to use promise 
    self.Permission.request(function () {
        return create_callback(title, options);
    });
} else {
    return create_callback(title, options);
}
```

It would be something like.

```js
// self.create is a Promise
return new Promise(function(resolve) {
    ...
    if (!self.Permission.has()) {
        // self.Permission.request is a promise
        retrurn self.Permission.request()
      .then(function() {
            // create_callback is also a promise 
            return create_callback(title, options);
        })
    } else {
        // create_callback is a promise
        return create_callback(title, options);
    }
    ...
}
```

**Comment by Nickersoft** [comment]
@sarbbottam I see what you're saying... in fact, when I posted earlier I overlooked the fact that that call is in-fact asynchronous and so doing some research I think promises would be appropriate. However, what is the browser compatibility like for this? According to caniuse.com, IE doesn't support promises at all. Is that a kind of shim framework that could be used for older browsers?

**Comment by sarbbottam** [comment]
>  According to caniuse.com, IE doesn't support promises at all. 

[IE doesn't support notifications either](http://caniuse.com/#search=notifications), I guess, it should be OK, for this use case. But its supported in [Edge](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#Browser_compatibility)

> Is that a kind of shim framework that could be used for older browsers?

[babel-polyfill](https://babeljs.io/docs/usage/polyfill/) could be used, but as stated I guess, it is not required.

**Comment by Nickersoft** [comment]
Wound up using the tiny 4kb library [promise-polyfill](https://github.com/taylorhakes/promise-polyfill) to fix this, as seen in PR #32. Closing issue.
